feat(http): optional query parameter to update only containers of a specified image

Foxite opened the pull request:
<!--

Thank you for contributing to the watchtower project! 🙏

We truly appreciate all the contributions we get from the community.

To make your PR experience as smooth as possible, make sure that you
include the following in your PR:

- What your PR contributes
- Which issues it solves (preferrably using auto closing instructions like "closes #123".
- Tests that verify the code your contributing
- Updates to the documentation

Thank you again! ✨

-->

Implements the second request from #1285: Added an optional query parameter that allows you to only update containers of the specified image. This works even if the image was built locally (ie the container no longer has the image label in `docker ps`).

Also includes an indentation fix that Goland went on strike over.

I will implement the first request later today, and probably add the ability to specify more than one image as well. This draft PR is so there can be a code review of what's already been done.

Don't hold back on roasting my Go skills. :P

Foxite committed:
feat(http): optional query parameter to update only containers of a specified image

Foxite committed:
fix style issues

Foxite committed:
comma separated image parameter

piksel commented on pkg/api/update/update.go:
```diff
@@ -43,12 +44,15 @@ func (handle *Handler) Handle(w http.ResponseWriter, r *http.Request) {
-		log.Debug("Skipped. Another update already running.")
+	var images []string
+	if r.URL.Query().Has("image") {
+		images = strings.Split(r.URL.Query().Get("image"), ",")
```
Since the URL query technically allows setting the same key multiple times, this would also allow for both the `?image=a&image=b` as well as the `?image=a,b` style:

```suggestion
	if imageQueries, found := r.URL.Query()["image"] {
		for _, image := range imageQueries {
			images = append(images, strings.Split(image, ","))
		}
```

If you only want to support the comma style, you could instead simply do:
```suggestion
	if image := r.URL.Query().Get("image"); image != "" {
		images = strings.Split(image, ",")
```

Foxite replied:
There was a compile error in here, I'm still not sure if this is because my Go version is incorrect, even though I've now explicitly told my IDE to use 1.15.

piksel commented on pkg/api/update/update.go:
```diff
@@ -43,12 +44,15 @@ func (handle *Handler) Handle(w http.ResponseWriter, r *http.Request) {
 
+	chanValue := <-lock
+	defer func() { lock <- chanValue }()
+	handle.fn(images)
```
This still need to check if the lock is taken before running the update, otherwise it's kind of pointless to have a lock...
```suggestion
select {
case chanValue := <-lock:
	defer func() { lock <- chanValue }()
	handle.fn(images)
default:
	log.Debug("Skipped. Another update already running.")
}
```

Foxite replied:
I could be wrong but I understood the original code to skip an update altogether when one is already running, while my change would make it wait until it's done before proceeding with the new update. Although I didn't actually look this up.

My idea was that you can issue multiple update requests in short succession, for different images (as a busy CI system might do), and they won't interfere with each other.

Foxite replied:
@piksel so my change makes the lock useless because it proceeds regardless of whether or not the lock was taken, but we should still allow multiple update requests for different images to be queued, no? I don't know how to do that, though, but I could find out of course.

piksel replied:
If the channel is sized to 1, it will block until the lock can be acquired. It does make more sense to queue up the requests now when they might have different parameters. I think it's generally fine, but since it's changed behaviour, maybe we can check if `len(images) > 0` and only block if that is the case. This is just to prevent any previously working setups from suddenly start checking continuously.

piksel replied:
If the channel is sized to 1, it will block until the lock can be acquired. It does make more sense to queue up the requests now when they might have different parameters. I think it's generally fine, but since it's changed behaviour, maybe we can check if `len(images) > 0` and only block if that is the case. This is just to prevent any previously working setups from suddenly start checking continuously.

piksel replied:
(I get that this was what your original code was doing, I just misunderstood what you meant with "useless" above. In my suggestion it should've been followed by a `return`)

piksel commented on cmd/root.go:
```diff
@@ -189,7 +189,7 @@ func Run(c *cobra.Command, names []string) {
 
 	if enableUpdateAPI {
-		updateHandler := update.New(func() { runUpdatesWithNotifications(filter) }, updateLock)
+		updateHandler := update.New(func(tags []string) { runUpdatesWithNotifications(filters.FilterByImageTag(tags, filter)) }, updateLock)
```
```suggestion
		updateHandler := update.New(func(images []string) { runUpdatesWithNotifications(filters.FilterByImageTag(images, filter)) }, updateLock)
```
We should be consistent, but perhaps even `imageTags` is the most opaque term to use...

Foxite replied:
So I understand that you think it should be changed to `imageTags` everywhere, is that correct?

Foxite replied:
I've spotted the difference between "image" and "tag"; I've changed the parameter/variable names to image.

piksel reviewed (changes requested):
In general, this looks good! The main blocker is the lock handling (and the URL.Values.Has call).

Foxite committed:
Support comma-separated query parameter as well as specifying it multiple times

Co-authored-by: nils måsén <nils@piksel.se>

Foxite committed:
fixed compile error

Foxite committed:
fixed FilterByImageTag

Not sure what changed in my testing setup, but Docker reports image names including the tag name now.

Foxite committed:
consistent use of image/tag (use image)

Foxite committed:
fixed multiple image queries

Foxite committed:
assuming I'm right here, only block on lock when any images are specified.

Foxite committed:
add unit tests for image filter. didn't add tests for update api because they didn't already exist

Foxite committed:
whoops.

Foxite committed:
use ImageName instead, add unit test for empty ImageName filter.

piksel reviewed (approved):
LGTM! Great job.
